crates/ruff/src/rules/flake8_todos/rules.rs

```diff
@@ -289,44 +288,34 @@ static ISSUE_LINK_REGEX_SET: Lazy<RegexSet> = Lazy::new(|| {
     .unwrap()
 });
 
-pub(crate) fn todos(tokens: &[LexResult], settings: &Settings) -> Vec<Diagnostic> {
+pub(crate) fn todos(indexer: &Indexer, locator: &Locator, settings: &Settings) -> Vec<Diagnostic> {
     let mut diagnostics: Vec<Diagnostic> = vec![];
 
-    let mut iter = tokens.iter().flatten().multipeek();
-    while let Some((token, token_range)) = iter.next() {
-        let Tok::Comment(comment) = token else {
-            continue;
-        };
+    let mut iter = indexer.comment_ranges().iter().multipeek();
+    while let Some(comment_range) = iter.next() {
+        let comment = locator.slice(*comment_range);
 
         // Check that the comment is a TODO (properly formed or not).
-        let Some(tag) = detect_tag(comment, token_range.start()) else {
+        let Some(tag) = detect_tag(comment, comment_range.start()) else {
             continue;
         };
 
         tag_errors(&tag, &mut diagnostics, settings);
-        static_errors(&mut diagnostics, comment, *token_range, &tag);
+        static_errors(&mut diagnostics, comment, *comment_range, &tag);
 
         // TD003
         let mut has_issue_link = false;
-        while let Some((token, token_range)) = iter.peek() {
-            match token {
-                Tok::Comment(comment) => {
-                    if detect_tag(comment, token_range.start()).is_some() {
-                        break;
-                    }
-                    if ISSUE_LINK_REGEX_SET.is_match(comment) {
-                        has_issue_link = true;
-                        break;
-                    }
-                }
-                Tok::Newline | Tok::NonLogicalNewline => {
-                    continue;
-                }
-                _ => {
-                    break;
-                }
+        while let Some(next_comment_range) = iter.peek() {
```

Comment:
Unrelated to this PR. Do we need multipeek here or could we rewrite this as 

```
while let Some(next_comment_range) = iter.peek() {
    let next_comment = locator.slice(*next_comment_range);
    if detect_tag(...).is_some() {
        break;
    }

    if ISSUE_LINK_REGEX_SET.is_matcch(next_comment) {
        has_issue_link = true;
        // Should we call `next()` here too because it is guaranteed that this comment has no tag
        // so it will fail the check on line 299
        break;
    }

    // skip over that comment because it neither has a tag nor link. It's not worth testing in the next iteration if it has a tag.
    iter.next();
}
```

Reply:
Yep, I think that's a good point! We really only care about `Tags`, so if we've determined that the `next_comment` doesn't have a tag, we can just skip it